File: javascript/src/billing/auto-recharge-service.ts
import Decimal from "decimal.js";
import { loadConfigFromDict } from "../config.js";
import {
  AutoRechargeDisabledError,
  AutoRechargeNotConfiguredError,
  PaymentMethodRequiredError,
  ProviderCapabilityNotSupportedError,
} from "../errors.js";
import type { BillingAutoRechargeProfile, BillingAutoRechargeStatus } from "./types/index.js";
import type {
  PaymentMethodInfo,
  PaymentProvider,
  SavedPaymentChargeQuote,
  SavedPaymentChargeResult,
} from "../providers/types.js";
import type { AutoRechargeBillingPort } from "./auto-recharge-port.js";
import { resolveAutoRechargeWindow } from "./policy-window.js";

export type AutoRechargeOutcome =
  | "not_configured"
  | "disabled"
  | "above_threshold"
  | "already_processing"
  | "limit_reached"
  | "submitted"
  | "action_required"
  | "failed";

export interface AutoRechargeProcessResult {
  outcome: AutoRechargeOutcome;
  charge?: SavedPaymentChargeResult | null;
}

interface ResolvedAutoRechargePolicy {
  enabled: true;
  threshold: Decimal;
  topupKey: string;
  topupId: string;
  quantity: number;
  maxChargesPerWindow: number;
  windowUnit: "second" | "minute" | "hour" | "day" | "week" | "month" | "year";
  windowCount: number;
  windowAnchor: "calendar" | "rolling";
  windowTimezone: string;
  windowStart: string;
  windowEnd: string;
  productId: string;
}

export class AutoRechargeService {
  constructor(private readonly billing: AutoRechargeBillingPort) {}

  private async policy(provider: PaymentProvider): Promise<ResolvedAutoRechargePolicy | null> {
    const raw = await this.billing.getActiveCatalogDocument();
    if (!raw) return null;
    const config = loadConfigFromDict(raw);
    const auto = config.commerce.autoRecharge;
    if (!auto) return null;

    const topupKey = auto.eligibleTopups[0];
    if (topupKey === undefined) return null;
    const topup = config.commerce.offers[topupKey];
    if (!topup || topup.type !== "topup") return null;
    const reference = topup.providers[provider.provider];
    if (!reference) return null;

    const productId =
      reference.type === "stripe_price"
        ? reference.priceId
        : reference.type === "dodo_product"
          ? reference.productId
          : reference.externalId;
    const resolvedTopup =
      reference.type === "stripe_price"
        ? await this.billing.resolveTopup(provider.provider, null, productId)
        : reference.type === "dodo_product"
          ? await this.billing.resolveTopup(provider.provider, productId, null)
          : await this.billing.resolveTopupByLookup(provider.provider, productId);
    if (!resolvedTopup) return null;

    const period = resolveAutoRechargeWindow(auto.limits.window);
    return {
      enabled: true,
      threshold: auto.balanceBelow.default,
      topupKey,
      topupId: resolvedTopup.topupId,
      quantity: auto.quantity.default,
      maxChargesPerWindow: auto.limits.maxPurchases,
      windowUnit: period.unit,
      windowCount: period.count,
      windowAnchor: period.anchor,
      windowTimezone: period.timezone,
      windowStart: period.start,
      windowEnd: period.end,
      productId,
    };
  }

  private async paymentMethod(
    userId: string,
    provider: PaymentProvider,
  ): Promise<{ customerId: string; method: PaymentMethodInfo } | null> {
    const customer = await this.billing.getCustomerByUserId(userId, provider.provider);
    if (!customer) return null;
    if (!provider.listPaymentMethods) {
      throw new ProviderCapabilityNotSupportedError(provider.provider, "listPaymentMethods");
    }
    const methods = await provider.listPaymentMethods(customer.providerCustomerId);
    const method =
      methods.find((candidate) => candidate.isDefault) ??
      (methods.length === 1 ? methods[0] : null);
    return method ? { customerId: customer.providerCustomerId, method } : null;
  }

  private preview(
    provider: PaymentProvider,
    policy: ResolvedAutoRechargePolicy,
    payment: { customerId: string; method: PaymentMethodInfo },
  ): Promise<SavedPaymentChargeQuote> | null {
    if (!provider.previewSavedPaymentCharge) return null;
    return provider.previewSavedPaymentCharge({
      customerId: payment.customerId,
      paymentMethodId: payment.method.id,
      productId: policy.productId,
      quantity: policy.quantity,
      metadata: {},
      idempotencyKey: "auto-recharge-preview",
    });
  }

  async quote(input: {
    userId: string;
    provider: PaymentProvider;
  }): Promise<SavedPaymentChargeQuote | null> {
    const policy = await this.policy(input.provider);
    if (!policy) return null;
    const payment = await this.paymentMethod(input.userId, input.provider);
    if (!payment) return null;
    return this.preview(input.provider, policy, payment);
  }

  async getStatus(input: {
    userId: string;
    provider: PaymentProvider;
  }): Promise<BillingAutoRechargeStatus | null> {
    const policy = await this.policy(input.provider);
    if (!policy) return null;
    const profile = await this.billing.getAutoRechargeProfile(input.userId);
    const payment = profile?.enabled
      ? await this.paymentMethod(input.userId, input.provider)
      : null;
    const quote = payment ? await this.preview(input.provider, policy, payment) : null;
    return {
      enabled: Boolean(profile?.enabled),
      state: profile?.enabled ? profile.state : "disabled",
      thresholdCredits: policy.threshold,
      topupKey: policy.topupKey,
      quantity: policy.quantity,
      maxRecharges: policy.maxChargesPerWindow,
      windowStart: policy.windowStart,
      windowEnd: policy.windowEnd,
      rechargesInWindow: await this.billing.countAutoRechargeAttempts(
        input.userId,
        policy.windowStart,
      ),
      paymentMethodId: payment?.method.id ?? null,
      paymentMethodLast4: payment?.method.last4 ?? null,
      paymentMethodBrand: payment?.method.brand ?? null,
      suspendedReason: profile?.state === "paused" ? "auto_recharge_paused" : null,
      pendingAttemptId: null,
      quoteAmountMinor: quote?.amountMinor ?? null,
      quoteCurrency: quote?.currency ?? null,
    };
  }

  async enable(input: {
    userId: string;
    provider: PaymentProvider;
    balance: Decimal;
    returnUrl?: string;
  }): Promise<BillingAutoRechargeStatus | null> {
    const policy = await this.policy(input.provider);
    if (!policy) throw new AutoRechargeNotConfiguredError();
    const payment = await this.paymentMethod(input.userId, input.provider);
    if (!payment) throw new PaymentMethodRequiredError();

    const profile: BillingAutoRechargeProfile = {
      userId: input.userId,
      enabled: true,
      state: "active",
      armed: true,
      provider: input.provider.provider,
      topupId: policy.topupId,
      quantity: policy.quantity,
      threshold: policy.threshold,
      maxChargesPerWindow: policy.maxChargesPerWindow,
      windowUnit: policy.windowUnit,
      windowCount: policy.windowCount,
      windowAnchor: policy.windowAnchor,
      windowTimezone: policy.windowTimezone,
    };
    await this.billing.upsertAutoRechargeProfile(profile);
    await this.processIfNeeded(input);
    return this.getStatus(input);
  }

  async disable(userId: string): Promise<void> {
    const profile = await this.billing.getAutoRechargeProfile(userId);
    if (!profile) return;
    await this.billing.upsertAutoRechargeProfile({
      ...profile,
      enabled: false,
      state: "disabled",
      armed: true,
    });
  }

  async retry(input: {
    userId: string;
    provider: PaymentProvider;
    balance: Decimal;
    returnUrl?: string;
  }): Promise<AutoRechargeProcessResult> {
    const profile = await this.billing.getAutoRechargeProfile(input.userId);
    if (!profile?.enabled) throw new AutoRechargeDisabledError();
    await this.billing.upsertAutoRechargeProfile(
      {
        ...profile,
        state: "active",
        armed: true,
      },
      { resetCooldown: true },
    );
    return this.processIfNeeded(input);
  }

  async processIfNeeded(input: {
    userId: string;
    provider: PaymentProvider;
    balance: Decimal;
    returnUrl?: string;
  }): Promise<AutoRechargeProcessResult> {
    const policy = await this.policy(input.provider);
    if (!policy) return { outcome: "not_configured" };
    const profile = await this.billing.getAutoRechargeProfile(input.userId);
    if (!profile?.enabled || profile.state !== "active") return { outcome: "disabled" };
    if (input.balance.gte(policy.threshold)) {
      return { outcome: "above_threshold" };
    }

    const payment = await this.paymentMethod(input.userId, input.provider);
    if (!payment) return { outcome: "failed" };
    if (!input.provider.chargeSavedPaymentMethod) {
      throw new ProviderCapabilityNotSupportedError(
        input.provider.provider,
        "chargeSavedPaymentMethod",
      );
    }
    const idempotencyKey = `auto-recharge:${input.userId}:${globalThis.crypto.randomUUID()}`;
    const attempt = await this.billing.claimAutoRechargeAttempt({
      userId: input.userId,
      idempotencyKey,
    });
    if (!attempt) return { outcome: "limit_reached" };
    // A transport error leaves the provider outcome unknown. The database
    // returns that in-flight attempt on the next claim; replay the provider
    // request with its original key so a committed charge cannot be duplicated.
    if (attempt.idempotencyKey !== idempotencyKey && attempt.state !== "unknown") {
      return { outcome: "already_processing" };
    }

    let charge: SavedPaymentChargeResult;
    try {
      charge = await input.provider.chargeSavedPaymentMethod({
        customerId: payment.customerId,
        paymentMethodId: payment.method.id,
        productId: policy.productId,
        quantity: policy.quantity,
        returnUrl: input.returnUrl,
        idempotencyKey: attempt.idempotencyKey,
        metadata: {
          auto_recharge_attempt_id: attempt.id,
          purpose: "credit_topup",
          bursar_account_id: input.userId,
        },
      });
    } catch (error) {
      await this.billing.updateAutoRechargeAttempt({
        id: attempt.id,
        state: "unknown",
        failureCode: "provider_request_failed",
        failureMessage: error instanceof Error ? error.message : String(error),
      });
      throw error;
    }
    if (charge.status === "requires_customer_action") {
      await this.billing.updateAutoRechargeAttempt({
        id: attempt.id,
        state: "action_required",
        providerAttemptId: charge.providerPaymentId ?? null,
      });
      return { outcome: "action_required", charge };
    }
    if (
      charge.status === "succeeded" ||
      charge.status === "processing" ||
      charge.status === "requires_merchant_action" ||
      charge.status === "requires_confirmation" ||
      charge.status === "requires_capture" ||
      charge.status === "partially_captured" ||
      charge.status === "partially_captured_and_capturable"
    ) {
      await this.billing.updateAutoRechargeAttempt({
        id: attempt.id,
        state: "processing",
        providerAttemptId: charge.providerPaymentId ?? null,
      });
      return { outcome: "submitted", charge };
    }

    await this.billing.updateAutoRechargeAttempt({
      id: attempt.id,
      state: "failed",
      providerAttemptId: charge.providerPaymentId ?? null,
      failureCode: "payment_failed",
    });
    return { outcome: "failed", charge };
  }
}

File: python/src/bursar/billing/auto_recharge_service.py
"""Provider-neutral auto-recharge orchestration."""

from __future__ import annotations

from dataclasses import dataclass
from decimal import Decimal
from typing import Any, Literal
from uuid import uuid4

from bursar.billing.policy_window import resolve_auto_recharge_window
from bursar.billing.types import (
    BillingAutoRechargeProfile,
    BillingAutoRechargeStatus,
    BillingTopupResult,
)
from bursar.config import (
    CustomObjectReference,
    DodoProductReference,
    StripePriceReference,
    TopupOffer,
    load_config_from_dict,
)
from bursar.providers.types import (
    PaymentMethodInfo,
    PaymentProvider,
    SavedPaymentChargeParams,
    SavedPaymentChargeQuote,
    SavedPaymentChargeResult,
)

AutoRechargeOutcome = Literal[
    "not_configured",
    "disabled",
    "above_threshold",
    "already_processing",
    "limit_reached",
    "submitted",
    "action_required",
    "failed",
]


@dataclass(frozen=True, slots=True)
class AutoRechargeProcessResult:
    outcome: AutoRechargeOutcome
    charge: SavedPaymentChargeResult | None = None


@dataclass(frozen=True, slots=True)
class _ResolvedAutoRechargePolicy:
    threshold: Decimal
    topup_key: str
    topup_id: str
    quantity: int
    max_charges_per_window: int
    window_unit: Literal["second", "minute", "hour", "day", "week", "month", "year"]
    window_count: int
    window_anchor: Literal["calendar", "rolling"]
    window_timezone: str
    window_start: str
    window_end: str
    window_days: float
    product_id: str


class AutoRechargeService:
    def __init__(self, billing: Any) -> None:
        self._billing = billing

    def _policy(self, provider: PaymentProvider) -> _ResolvedAutoRechargePolicy | None:
        raw = self._billing.get_active_bursar_config()
        if raw is None:
            return None
        config = load_config_from_dict(raw)
        policy = config.commerce.auto_recharge
        if policy is None:
            return None

        topup_key = policy.eligible_topups[0]
        topup = config.commerce.offers.get(topup_key)
        if not isinstance(topup, TopupOffer):
            return None
        reference = topup.providers.get(provider.provider)
        if reference is None:
            return None

        resolved_topup: BillingTopupResult | None
        if isinstance(reference, StripePriceReference):
            product_id = reference.price_id
            resolved_topup = self._billing.resolve_topup(provider.provider, None, product_id)
        elif isinstance(reference, DodoProductReference):
            product_id = reference.product_id
            resolved_topup = self._billing.resolve_topup(provider.provider, product_id, None)
        elif isinstance(reference, CustomObjectReference):
            product_id = reference.external_id
            resolved_topup = self._billing.resolve_topup_by_lookup(provider.provider, product_id)
        else:
            return None
        if resolved_topup is None:
            return None

        period = resolve_auto_recharge_window(policy.limits.window)
        return _ResolvedAutoRechargePolicy(
            threshold=Decimal(policy.balance_below.default),
            topup_key=topup_key,
            topup_id=resolved_topup.topup_id,
            quantity=policy.quantity.default,
            max_charges_per_window=policy.limits.max_purchases,
            window_unit=period.unit,
            window_count=period.count,
            window_anchor=period.anchor,
            window_timezone=period.timezone,
            window_start=period.start,
            window_end=period.end,
            window_days=period.duration_days,
            product_id=product_id,
        )

    async def _payment_method(
        self,
        user_id: str,
        provider: PaymentProvider,
    ) -> tuple[str, PaymentMethodInfo] | None:
        customer = self._billing.get_customer_by_user_id(user_id, provider.provider)
        if customer is None:
            return None
        methods = await provider.list_payment_methods(customer.provider_customer_id)
        method = await provider.get_default_payment_method(customer.provider_customer_id)
        if method is None:
            method = next((candidate for candidate in methods if candidate.is_default), None)
        if method is None and len(methods) == 1:
            method = methods[0]
        return (customer.provider_customer_id, method) if method is not None else None

    async def quote(
        self,
        user_id: str,
        provider: PaymentProvider,
    ) -> SavedPaymentChargeQuote | None:
        policy = self._policy(provider)
        if policy is None:
            return None
        payment = await self._payment_method(user_id, provider)
        if payment is None:
            return None
        customer_id, method = payment
        try:
            return await provider.preview_saved_payment_charge(
                SavedPaymentChargeParams(
                    customer_id=customer_id,
                    payment_method_id=method.id,
                    product_id=policy.product_id,
                    quantity=policy.quantity,
                    metadata={},
                    idempotency_key="auto-recharge-preview",
                )
            )
        except NotImplementedError:
            return None

    async def get_status(
        self,
        user_id: str,
        provider: PaymentProvider,
    ) -> BillingAutoRechargeStatus | None:
        policy = self._policy(provider)
        if policy is None:
            return None
        profile = self._billing.get_auto_recharge_profile(user_id)
        payment = await self._payment_method(user_id, provider) if profile is not None and profile.enabled else None
        quote = await self.quote(user_id, provider)
        method = payment[1] if payment is not None else None
        return BillingAutoRechargeStatus(
            enabled=bool(profile and profile.enabled),
            state=profile.state if profile and profile.enabled else "disabled",
            threshold_credits=policy.threshold,
            topup_key=policy.topup_key,
            quantity=policy.quantity,
            max_recharges=policy.max_charges_per_window,
            window_days=policy.window_days,
            window_start=policy.window_start,
            window_end=policy.window_end,
            recharges_in_window=self._billing.count_auto_recharge_attempts(
                user_id,
                policy.window_start,
            ),
            payment_method_id=method.id if method is not None else None,
            payment_method_last4=method.last4 if method is not None else None,
            payment_method_brand=method.brand if method is not None else None,
            suspended_reason=("auto_recharge_paused" if profile and profile.state == "paused" else None),
            pending_attempt_id=None,
            quote_amount_minor=quote.amount_minor if quote is not None else None,
            quote_currency=quote.currency if quote is not None else None,
        )

    async def enable(
        self,
        user_id: str,
        provider: PaymentProvider,
        *,
        balance: Decimal | int,
        return_url: str,
        consent_reference: str | None = None,
    ) -> BillingAutoRechargeStatus | None:
        policy = self._policy(provider)
        if policy is None:
            raise ValueError("auto_recharge_not_configured")
        payment = await self._payment_method(user_id, provider)
        if payment is None:
            raise ValueError("payment_method_required")

        self._billing.upsert_auto_recharge_profile(
            BillingAutoRechargeProfile(
                user_id=user_id,
                enabled=True,
                state="active",
                armed=True,
                provider=provider.provider,
                topup_id=policy.topup_id,
                quantity=policy.quantity,
                threshold=policy.threshold,
                max_charges_per_window=policy.max_charges_per_window,
                window_unit=policy.window_unit,
                window_count=policy.window_count,
                window_anchor=policy.window_anchor,
                window_timezone=policy.window_timezone,
            )
        )
        await self.process_if_needed(
            user_id,
            provider,
            balance=balance,
            return_url=return_url,
        )
        return await self.get_status(user_id, provider)

    def disable(self, user_id: str) -> None:
        profile = self._billing.get_auto_recharge_profile(user_id)
        if profile is None:
            return
        self._billing.upsert_auto_recharge_profile(
            profile.model_copy(
                update={
                    "enabled": False,
                    "state": "disabled",
                    "armed": True,
                }
            )
        )

    async def retry(
        self,
        user_id: str,
        provider: PaymentProvider,
        *,
        balance: Decimal | int,
        return_url: str,
    ) -> AutoRechargeProcessResult:
        profile = self._billing.get_auto_recharge_profile(user_id)
        if profile is None or not profile.enabled:
            raise ValueError("auto_recharge_disabled")
        self._billing.upsert_auto_recharge_profile(profile.model_copy(update={"state": "active", "armed": True}))
        return await self.process_if_needed(
            user_id,
            provider,
            balance=balance,
            return_url=return_url,
        )

    async def process_if_needed(
        self,
        user_id: str,
        provider: PaymentProvider,
        *,
        balance: Decimal | int,
        return_url: str,
    ) -> AutoRechargeProcessResult:
        policy = self._policy(provider)
        if policy is None:
            return AutoRechargeProcessResult(outcome="not_configured")
        profile = self._billing.get_auto_recharge_profile(user_id)
        if profile is None or not profile.enabled or profile.state != "active":
            return AutoRechargeProcessResult(outcome="disabled")
        if Decimal(balance) >= policy.threshold:
            if not profile.armed:
                self._billing.upsert_auto_recharge_profile(profile.model_copy(update={"armed": True}))
            return AutoRechargeProcessResult(outcome="above_threshold")

        payment = await self._payment_method(user_id, provider)
        if payment is None:
            return AutoRechargeProcessResult(outcome="failed")
        customer_id, method = payment
        attempt = self._billing.claim_auto_recharge_attempt(
            user_id,
            f"auto-recharge:{user_id}:{uuid4()}",
        )
        if attempt is None:
            return AutoRechargeProcessResult(outcome="limit_reached")

        charge = await provider.charge_saved_payment_method(
            SavedPaymentChargeParams(
                customer_id=customer_id,
                payment_method_id=method.id,
                product_id=policy.product_id,
                quantity=policy.quantity,
                return_url=return_url,
                idempotency_key=attempt.idempotency_key,
                metadata={
                    "auto_recharge_attempt_id": attempt.id,
                    "purpose": "credit_topup",
                    "userId": user_id,
                },
            )
        )
        if charge.status == "requires_customer_action":
            self._billing.update_auto_recharge_attempt(
                attempt.id,
                "submitted",
                charge.provider_payment_id,
            )
            self._billing.upsert_auto_recharge_profile(profile.model_copy(update={"state": "paused"}))
            return AutoRechargeProcessResult(outcome="action_required", charge=charge)

        if charge.status in {"succeeded", "processing"}:
            self._billing.update_auto_recharge_attempt(
                attempt.id,
                "processing",
                charge.provider_payment_id,
            )
            return AutoRechargeProcessResult(outcome="submitted", charge=charge)

        self._billing.update_auto_recharge_attempt(
            attempt.id,
            "failed",
            charge.provider_payment_id,
            "payment_failed",
        )
        self._billing.upsert_auto_recharge_profile(profile.model_copy(update={"state": "paused"}))
        return AutoRechargeProcessResult(outcome="failed", charge=charge)
